File: demoPy/productor_pedidos.py
import pika
import json
from datetime import datetime

"""
Productor de Pedidos E-commerce - Python
=========================================
Caso de uso: sistema de e-commerce que enruta pedidos a diferentes
colas segun el tipo de envio mediante un exchange de tipo 'topic'.

Routing keys:
  pedido.express       -> entrega en 24h
  pedido.normal        -> entrega en 3-5 dias
  pedido.internacional -> entrega en 15-30 dias
"""

EXCHANGE_NAME = "pedidos_exchange"

def enviar_pedido(channel, routing_key, id_pedido, producto, cantidad, precio):
    """
    Publica un pedido en el exchange con la routing key indicada.
    El mensaje se serializa como JSON para que sea facil de parsear por el consumidor.
    """
    pedido = {
        "id": id_pedido,
        "producto": producto,
        "cantidad": cantidad,
        "precio": precio,
        "timestamp": datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    }
    # Serializar el pedido a JSON y codificarlo en bytes
    body = json.dumps(pedido).encode("utf-8")

    channel.basic_publish(
        exchange=EXCHANGE_NAME,
        routing_key=routing_key,
        body=body
    )
    print(f"[Productor] Pedido enviado [{routing_key}] -> {json.dumps(pedido)}")


def main():
    # 1. Conectar a RabbitMQ
    connection = pika.BlockingConnection(pika.ConnectionParameters("localhost"))
    channel = connection.channel()

    # 2. Declarar el exchange de tipo topic
    channel.exchange_declare(exchange=EXCHANGE_NAME, exchange_type="topic")

    # 3. Enviar pedidos de diferentes tipos
    enviar_pedido(channel, "pedido.express",       "2001", "Laptop Gamer",        2, 3500000)
    enviar_pedido(channel, "pedido.normal",        "2002", "Teclado Mecanico",    1,  250000)
    enviar_pedido(channel, "pedido.internacional", "2003", "Camara DSLR",         1, 4200000)
    enviar_pedido(channel, "pedido.express",       "2004", "Audifonos Bluetooth", 3,  180000)
    enviar_pedido(channel, "pedido.normal",        "2005", "Mouse Gamer",         2,  120000)

    print("\n[Productor] Todos los pedidos han sido enviados.")
    connection.close()


if __name__ == "__main__":
    main()

File: demo/src/main/java/com/example/ProductorPedidos.java
package com.example;

import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.ConnectionFactory;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

/**
 * Productor de Pedidos E-commerce
 *
 * Caso de uso: Un sistema de e-commerce recibe pedidos y los enruta
 * a diferentes colas segun el tipo de envio usando un exchange de tipo "topic".
 *
 * Routing keys usadas:
 *   pedido.express       -> consumidor de pedidos express (envio en 24h)
 *   pedido.normal        -> consumidor de pedidos normales (envio en 3-5 dias)
 *   pedido.internacional -> consumidor de pedidos internacionales (envio en 15-30 dias)
 */
public class ProductorPedidos {

    // Nombre del exchange topic
    private static final String EXCHANGE_NAME = "pedidos_exchange";

    public static void main(String[] args) throws Exception {

        // 1. Crear la fabrica de conexiones y conectar a RabbitMQ local
        ConnectionFactory factory = new ConnectionFactory();
        factory.setHost("localhost");
        factory.setPort(5672);

        try (Connection connection = factory.newConnection();
             Channel channel = connection.createChannel()) {

            // 2. Declarar el exchange de tipo "topic"
            //    topic permite filtrar mensajes por patron de routing key (e.g. pedido.*)
            channel.exchangeDeclare(EXCHANGE_NAME, "topic");

            // 3. Simular el envio de pedidos de distintos tipos
            enviarPedido(channel, "pedido.express", "1001", "Laptop Gamer", 2, 3500000);
            enviarPedido(channel, "pedido.normal",  "1002", "Teclado Mecanico", 1, 250000);
            enviarPedido(channel, "pedido.internacional", "1003", "Camara DSLR", 1, 4200000);
            enviarPedido(channel, "pedido.express", "1004", "Audifonos Bluetooth", 3, 180000);
            enviarPedido(channel, "pedido.normal",  "1005", "Mouse Gamer", 2, 120000);

            System.out.println("\n[Productor] Todos los pedidos han sido enviados a la cola.");
        }
    }

    /**
     * Construye y publica un mensaje de pedido en el exchange.
     *
     * @param channel      Canal RabbitMQ activo
     * @param routingKey   Clave de enrutamiento (define el tipo de pedido)
     * @param idPedido     Identificador unico del pedido
     * @param producto     Nombre del producto
     * @param cantidad     Cantidad solicitada
     * @param precio       Precio unitario en pesos colombianos
     */
    private static void enviarPedido(Channel channel, String routingKey,
                                     String idPedido, String producto,
                                     int cantidad, long precio) throws Exception {

        // Timestamp del momento en que se genera el pedido
        String timestamp = LocalDateTime.now()
                .format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));

        // Construir el mensaje como texto simple (JSON manual para no depender de librerias extra)
        String mensaje = String.format(
            "{\"id\":\"%s\",\"producto\":\"%s\",\"cantidad\":%d,\"precio\":%d,\"timestamp\":\"%s\"}",
            idPedido, producto, cantidad, precio, timestamp
        );

        // Publicar el mensaje en el exchange con la routing key correspondiente
        channel.basicPublish(EXCHANGE_NAME, routingKey, null, mensaje.getBytes("UTF-8"));

        System.out.printf("[Productor] Pedido enviado [%s] -> %s%n", routingKey, mensaje);
    }
}
